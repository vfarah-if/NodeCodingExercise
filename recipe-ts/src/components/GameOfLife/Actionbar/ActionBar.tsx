import React from 'react'
import Button from '../../Button'
import './style/index.css'

export interface ActionbarProps {
  title?: string
  generateDisplay?: string
  simulateDisplay?: string
  randomiseDisplay?: string
  clearDisplay?: string
  onGenerate: () => void
  onSimulate: () => void
  onRandomise: () => void
  onClear: () => void
}

const Actionbar: React.FC<ActionbarProps> = ({
  title = 'Game of Life on Mars',
  generateDisplay = 'Generate',
  simulateDisplay = 'Simulate',
  randomiseDisplay = 'Randomise',
  clearDisplay = 'Clear',
  onGenerate,
  onSimulate,
  onRandomise,
  onClear,
}) => {
  return (
    <header>
      <div className='wrapper'>
        <div>
          <svg
            version='1.0'
            xmlns='http://www.w3.org/2000/svg'
            xmlnsXlink='http://www.w3.org/1999/xlink'
            width='72'
            height='58'
          >
            <defs>
              <rect
                id='alive'
                x='0'
                y='0'
                width='14'
                height='14'
                fill='black'
                stroke='#c0c0c0'
                strokeWidth='2'
              />
              <rect
                id='dead'
                x='0'
                y='0'
                width='14'
                height='14'
                fill='white'
                stroke='#c0c0c0'
                strokeWidth='2'
              />
            </defs>
            <g transform='translate(1,1)'>
              <use xlinkHref='#dead' transform='translate(0,0)' />
              <use xlinkHref='#alive' transform='translate(14,0)' />

              <use xlinkHref='#dead' transform='translate(28,0)' />
              <use xlinkHref='#dead' transform='translate(42,0)' />
              <use xlinkHref='#alive' transform='translate(56,0)' />

              <use xlinkHref='#alive' transform='translate(0,14)' />
              <use xlinkHref='#dead' transform='translate(14,14)' />
              <use xlinkHref='#dead' transform='translate(28,14)' />
              <use xlinkHref='#dead' transform='translate(42,14)' />
              <use xlinkHref='#dead' transform='translate(56,14)' />

              <use xlinkHref='#alive' transform='translate(0,28)' />
              <use xlinkHref='#dead' transform='translate(14,28)' />
              <use xlinkHref='#dead' transform='translate(28,28)' />
              <use xlinkHref='#dead' transform='translate(42,28)' />
              <use xlinkHref='#alive' transform='translate(56,28)' />

              <use xlinkHref='#alive' transform='translate(0,42)' />
              <use xlinkHref='#alive' transform='translate(14,42)' />
              <use xlinkHref='#alive' transform='translate(28,42)' />

              <use xlinkHref='#alive' transform='translate(42,42)' />
              <use xlinkHref='#dead' transform='translate(56,42)' />
            </g>
          </svg>
          <h1>{title}</h1>
        </div>
        <div>
          <Button
            primary
            size='small'
            onClick={onGenerate}
            label={generateDisplay}
          />
          <Button
            primary
            size='small'
            onClick={onSimulate}
            label={simulateDisplay}
          />
          <Button
            primary
            size='small'
            onClick={onRandomise}
            label={randomiseDisplay}
          ></Button>
          <Button
            primary
            size='small'
            onClick={onClear}
            label={clearDisplay}
          ></Button>
        </div>
      </div>
    </header>
  )
}

export default Actionbar
