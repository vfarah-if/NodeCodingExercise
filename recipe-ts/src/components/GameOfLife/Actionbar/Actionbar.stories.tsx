import { action } from '@storybook/addon-actions'
import { Meta, Story } from '@storybook/react/types-6-0'

import Actionbar, { ActionbarProps } from './ActionBar'

export default {
  title: 'Component/GameOfLife/Actionbar',
  component: Actionbar,
} as Meta

const ActionbarTemplate: Story<ActionbarProps> = args => (
  <>
    <Actionbar
      {...args}
      onClear={action('onClear...')}
      onGenerate={action('onGenerate...')}
      onRandomise={action('onRandomise...')}
      onSimulate={action('onSimulate...')}
    />
  </>
)

export const DefaultActionbar = ActionbarTemplate.bind({})
DefaultActionbar.args = {}

export const CustomActionbar = ActionbarTemplate.bind({})
CustomActionbar.args = {
  title: "Conway's Game of life",
  clearDisplay: 'Erase',
  generateDisplay: 'Make it so',
  simulateDisplay: 'Animate',
  randomiseDisplay: 'Shuffle it',
}
