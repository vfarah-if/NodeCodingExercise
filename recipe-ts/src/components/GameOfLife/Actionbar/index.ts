import Actionbar from './ActionBar'
export default Actionbar
