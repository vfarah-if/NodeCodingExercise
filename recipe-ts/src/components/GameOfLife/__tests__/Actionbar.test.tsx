import { render, screen } from '@testing-library/react';
import Actionbar from '../Actionbar';

import {
  DefaultActionbar,
  CustomActionbar,
} from '../Actionbar/Actionbar.stories';

describe('Actionbar', () => {
  const generateHandler = jest.fn();
  const clearHandler = jest.fn();
  const randomiseHandler = jest.fn();
  const simulateHandler = jest.fn();

  beforeEach(() => {
    generateHandler.mockClear();
    clearHandler.mockClear();
    randomiseHandler.mockClear();
    simulateHandler.mockClear();
  });

  test('should default expected action settings', () => {
    const { container } = render(
      <DefaultActionbar
        {...DefaultActionbar.args}
        onClear={clearHandler}
        onGenerate={generateHandler}
        onRandomise={randomiseHandler}
        onSimulate={simulateHandler}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should customise action bar settings', () => {
    const { container } = render(
      <CustomActionbar
        {...CustomActionbar.args}
        onClear={clearHandler}
        onGenerate={generateHandler}
        onRandomise={randomiseHandler}
        onSimulate={simulateHandler}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should command clear event', () => {
    render(
      <Actionbar
        {...CustomActionbar.args}
        onClear={clearHandler}
        onGenerate={generateHandler}
        onRandomise={randomiseHandler}
        onSimulate={simulateHandler}
      />
    );

    const clearButton = screen.getByText('Erase');
    clearButton.click();

    expect(clearHandler).toHaveBeenCalledTimes(1);
  });

  test('should command generate event', () => {
    render(
      <Actionbar
        {...CustomActionbar.args}
        onClear={clearHandler}
        onGenerate={generateHandler}
        onRandomise={randomiseHandler}
        onSimulate={simulateHandler}
      />
    );
    const generateButton = screen.getByText('Make it so');

    generateButton.click();

    expect(generateHandler).toHaveBeenCalledTimes(1);
  });

  test('should command randomise event', () => {
    render(
      <Actionbar
        {...CustomActionbar.args}
        onClear={clearHandler}
        onGenerate={generateHandler}
        onRandomise={randomiseHandler}
        onSimulate={simulateHandler}
      />
    );
    const randomiseButton = screen.getByText('Shuffle it');

    randomiseButton.click();

    expect(randomiseHandler).toHaveBeenCalledTimes(1);
  });

  test('should command simulate event', () => {
    render(
      <Actionbar
        {...CustomActionbar.args}
        onClear={clearHandler}
        onGenerate={generateHandler}
        onRandomise={randomiseHandler}
        onSimulate={simulateHandler}
      />
    );
    const simulateButton = screen.getByText('Animate');

    simulateButton.click();

    expect(simulateHandler).toHaveBeenCalledTimes(1);
  });
});
