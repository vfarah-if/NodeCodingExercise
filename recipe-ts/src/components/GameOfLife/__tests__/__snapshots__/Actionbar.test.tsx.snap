// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actionbar should customise action bar settings 1`] = `
<div>
  <header>
    <div
      class="wrapper"
    >
      <div>
        <svg
          height="58"
          version="1.0"
          width="72"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <defs>
            <rect
              fill="black"
              height="14"
              id="alive"
              stroke="#c0c0c0"
              stroke-width="2"
              width="14"
              x="0"
              y="0"
            />
            <rect
              fill="white"
              height="14"
              id="dead"
              stroke="#c0c0c0"
              stroke-width="2"
              width="14"
              x="0"
              y="0"
            />
          </defs>
          <g
            transform="translate(1,1)"
          >
            <use
              transform="translate(0,0)"
              xlink:href="#dead"
            />
            <use
              transform="translate(14,0)"
              xlink:href="#alive"
            />
            <use
              transform="translate(28,0)"
              xlink:href="#dead"
            />
            <use
              transform="translate(42,0)"
              xlink:href="#dead"
            />
            <use
              transform="translate(56,0)"
              xlink:href="#alive"
            />
            <use
              transform="translate(0,14)"
              xlink:href="#alive"
            />
            <use
              transform="translate(14,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(28,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(42,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(56,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(0,28)"
              xlink:href="#alive"
            />
            <use
              transform="translate(14,28)"
              xlink:href="#dead"
            />
            <use
              transform="translate(28,28)"
              xlink:href="#dead"
            />
            <use
              transform="translate(42,28)"
              xlink:href="#dead"
            />
            <use
              transform="translate(56,28)"
              xlink:href="#alive"
            />
            <use
              transform="translate(0,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(14,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(28,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(42,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(56,42)"
              xlink:href="#dead"
            />
          </g>
        </svg>
        <h1>
          Conway's Game of life
        </h1>
      </div>
      <div>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Make it so"
        >
          Make it so
        </button>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Animate"
        >
          Animate
        </button>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Shuffle it"
        >
          Shuffle it
        </button>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Erase"
        >
          Erase
        </button>
      </div>
    </div>
  </header>
</div>
`;

exports[`Actionbar should default expected action settings 1`] = `
<div>
  <header>
    <div
      class="wrapper"
    >
      <div>
        <svg
          height="58"
          version="1.0"
          width="72"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <defs>
            <rect
              fill="black"
              height="14"
              id="alive"
              stroke="#c0c0c0"
              stroke-width="2"
              width="14"
              x="0"
              y="0"
            />
            <rect
              fill="white"
              height="14"
              id="dead"
              stroke="#c0c0c0"
              stroke-width="2"
              width="14"
              x="0"
              y="0"
            />
          </defs>
          <g
            transform="translate(1,1)"
          >
            <use
              transform="translate(0,0)"
              xlink:href="#dead"
            />
            <use
              transform="translate(14,0)"
              xlink:href="#alive"
            />
            <use
              transform="translate(28,0)"
              xlink:href="#dead"
            />
            <use
              transform="translate(42,0)"
              xlink:href="#dead"
            />
            <use
              transform="translate(56,0)"
              xlink:href="#alive"
            />
            <use
              transform="translate(0,14)"
              xlink:href="#alive"
            />
            <use
              transform="translate(14,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(28,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(42,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(56,14)"
              xlink:href="#dead"
            />
            <use
              transform="translate(0,28)"
              xlink:href="#alive"
            />
            <use
              transform="translate(14,28)"
              xlink:href="#dead"
            />
            <use
              transform="translate(28,28)"
              xlink:href="#dead"
            />
            <use
              transform="translate(42,28)"
              xlink:href="#dead"
            />
            <use
              transform="translate(56,28)"
              xlink:href="#alive"
            />
            <use
              transform="translate(0,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(14,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(28,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(42,42)"
              xlink:href="#alive"
            />
            <use
              transform="translate(56,42)"
              xlink:href="#dead"
            />
          </g>
        </svg>
        <h1>
          Game of Life on Mars
        </h1>
      </div>
      <div>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Generate"
        >
          Generate
        </button>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Simulate"
        >
          Simulate
        </button>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Randomise"
        >
          Randomise
        </button>
        <button
          class="sc-bcXHqe dkxJLe"
          label="Clear"
        >
          Clear
        </button>
      </div>
    </div>
  </header>
</div>
`;
